components/suggest/src/yelp.rs

```diff
@@ -296,6 +301,46 @@ impl<'a> SuggestDao<'a> {
         ))
     }
 
+    /// Find the subject from the given string.
+    /// It returns the Option. If it is not none, it contains the tuple as follows:
+    /// (
+    ///   String: Subject.
+    ///   bool: Whether the subject matched exactly with the paramter.
+    /// )
+    fn find_subject(&self, candidate: &str) -> Result<Option<(String, bool)>> {
+        if candidate.is_empty() {
+            return Ok(None);
+        }
+
+        // If the length of subject candidate is less than PREFIX_MATCH_THRESHOLD,
+        // should exact match.
+        if candidate.len() < PREFIX_MATCH_THRESHOLD {
+            return Ok(if self.is_subject(candidate)? {
+                Some((candidate.to_string(), true))
+            } else {
+                None
+            });
+        }
+
+        // Otherwise, apply prefix-match.
+        Ok(
+            match self.conn.query_row_and_then_cachable(
+                "SELECT keyword FROM yelp_subjects WHERE keyword LIKE :candidate || '%' ORDER BY LENGTH(keyword) ASC LIMIT 1",
```

Comment:
Small suggestion: Using `BETWEEN` instead of `LIKE` will let SQLite take advantage of the automatic index on the primary key, instead of doing a full table scan:

```suggestion
                "SELECT keyword FROM yelp_subjects WHERE keyword BETWEEN :candidate AND :candidate || x'FFFF' ORDER BY LENGTH(keyword) ASC LIMIT 1",
```

The optimizer [can optimize `x%` `LIKE` patterns to be prefix matches](https://www.sqlite.org/optoverview.html#the_like_optimization), but won't by default—because `TEXT` columns are case-sensitive unless they're explicitly declared `COLLATE NOCASE`, but `LIKE` is case-insensitive by default.

The `ORDER BY LENGTH(keyword)` will still need to create a temporary sorting B-Tree, but it should only have a small handful of entries.

Reply:
Oh, I see, thanks!

Reply:
Nice catch, @linabutler! TIL💡